cider/tests/utils/QueryArrowDataGenerator.h

```diff
@@ -0,0 +1,146 @@
+    builder = builder.addColumn<C_TYPE>(names[i], type, col_data, null_data); \
+    break;                                                                    \
+  }
+
+#define N_MAX std::numeric_limits<T>::max()
+
+#define N_MIN std::numeric_limits<T>::min()
+
+class QueryArrowDataGenerator {
+ public:
+  static void generateBatchByTypes(ArrowSchema*& schema,
+                                   ArrowArray*& array,
+                                   const size_t row_num,
+                                   const std::vector<std::string>& names,
+                                   const std::vector<::substrait::Type>& types,
+                                   std::vector<int32_t> null_chance = {},
+                                   GeneratePattern pattern = GeneratePattern::Sequence,
+                                   const int64_t value_min = 0,
+                                   const int64_t value_max = -1) {
+    if (null_chance.empty()) {
+      null_chance = std::vector<int32_t>(types.size(), 0);
+    }
+    ArrowArrayBuilder builder;
+    builder = builder.setRowNum(row_num);
+    for (auto i = 0; i < types.size(); ++i) {
+      ::substrait::Type type = types[i];
+      switch (type.kind_case()) {
+        case ::substrait::Type::KindCase::kI8:
+          GENERATE_AND_ADD_COLUMN(int8_t)
+        case ::substrait::Type::KindCase::kI16:
+          GENERATE_AND_ADD_COLUMN(int16_t)
+        case ::substrait::Type::KindCase::kI32:
+          GENERATE_AND_ADD_COLUMN(int32_t)
+        case ::substrait::Type::KindCase::kI64:
+          GENERATE_AND_ADD_COLUMN(int64_t)
+        case ::substrait::Type::KindCase::kFp32:
+          GENERATE_AND_ADD_COLUMN(float)
+        case ::substrait::Type::KindCase::kFp64:
+          GENERATE_AND_ADD_COLUMN(double)
+        default:
+          CIDER_THROW(CiderCompileException, "Type not supported.");
+      }
+    }
+
+    auto schema_array = builder.build();
+    schema = std::get<0>(schema_array);
+    array = std::get<1>(schema_array);
+  }
+
+ private:
+  template <typename T>
+  static std::tuple<std::vector<T>, std::vector<bool>> generateAndFillVector(
+      const size_t row_num,
+      const GeneratePattern pattern,
+      const int32_t
+          null_chance,  // Null chance for each column, -1 represents for unnullable
+                        // column, 0 represents for nullable column but all data is not
+                        // null, 1 represents for all rows are null, values >= 2 means
+                        // each row has 1/x chance to be null.
+      const T value_min = N_MIN,
+      const T value_max = N_MAX) {
+    std::vector<T> col_data(row_num);
+    std::vector<bool> null_data(row_num);
+    std::mt19937 rng(std::random_device{}());  // NOLINT
+    switch (pattern) {
+      case GeneratePattern::Sequence:
+        for (auto i = 0; i < row_num; ++i) {
+          null_data[i] = Random::oneIn(null_chance, rng) ? (col_data[i] = N_MIN, true)
```

Comment:
we don't need this `col_data[i] = N_MIN` anymore. could use 0

Reply:
do you mean replace "N_MIN" with "0" here?

Reply:
yes `col_data[i] = 0` would be fine.

Reply:
We may need to verify a scenario that the value is actual ` N_MIN` but in null_vector, it's set to false, which means it's not a null value and should be selected. Should we add some logic here?